### PR TITLE
tidb: update statement history in retry()

### PR DESCRIPTION
**Pull request by tiancaiamao**
Fix a bug in this order:
1. transaction commit meet schema change
2. retry and meet retryable error
3. retry again and success

It would insert stale data, because statement history is not updated,
we may using the wrong schema.

@shenli @coocood @zimulala

### CODE REVIEW

**Comment by shenli** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Comment by zimulala** [comment]
Waiting for tests now.

File: session.go [hunk 1]
```diff
@@ -448,6 +449,10 @@ func (s *session) retry(goCtx goctx.Context, maxCnt int, infoSchemaChanged bool)
 				break
 			}
 		}
+		if ctx.Value("mockRetryForTest") != nil {
+			err = kv.ErrRetryable
+			ctx = goctx.Background()
```
Comment (shenli) [inline session.go]: Why create a new ctx?

File: session.go [hunk 2]
```diff
@@ -449,6 +449,10 @@ func (s *session) retry(goCtx goctx.Context, maxCnt int, infoSchemaChanged bool)
 				break
 			}
 		}
+		if ctx.Value("mockRetryForTest") != nil {
```
Comment (coocood) [inline session.go]: We don't need to do this.

File: session.go (hunk 1 above)
Comment (tiancaiamao) [reply session.go]: If we don't reset the ctx, the mock code will run to here all the time and dead loop...
Here it means set a mock flag, do something, clear the flag as we've done what we need.

File: new_session_test.go [hunk 3]
```diff
@@ -1461,12 +1461,31 @@ func (s *testSchemaSuite) TestCommitWhenSchemaChanged(c *C) {
...
+	tk1.MustExec("insert into t values (1)")
+
+	tk.MustExec("alter table t add column (b int)")
+
+	// When tk1 commit, it will find schema already changed.
+	// mockRetryForTest would make retry() fail for one more time,
+	// and retry again, to cover a bug that statement history is not
+	// updated during retry.
+	// See https://github.com/pingcap/tidb/pull/5202
+	err := tk1.Se.CommitTxn(goctx.WithValue(goctx.Background(), "mockRetryForTest", 1))
```
Comment (coocood) [inline new_session_test.go]: We use another session to make a write conflict.

File: session.go (hunk 2 above)
Comment (tiancaiamao) [reply session.go]: It's ugly but how do you test the case?

File: new_session_test.go (hunk 3 above)
Comment (tiancaiamao) [reply new_session_test.go]: Good idea, but the logic is not under control after commit is called.

File: session.go (hunk 2 above)
Comment (zimulala) [reply session.go]: Can we add a hook here in the future? It can be used to add more flexible tests？

Comment (tiancaiamao) [reply session.go]:
```
func retry(ctx) {
   hook := ctx.Value("mockRetryForTest")
   doActionCheckHook()
}

ctx := goctx.WithValue("mockRetryForTest", Hook)
retry(ctx)
hook.Status = 1
hook.Status = 2
hook.Status = 3
```
@zimulala

**Comment by shenli** [comment]
@coocood PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**